dask/array/tests/test_distributed.py

```diff
@@ -0,0 +1,33 @@
+import pytest
+
+pytest.importorskip("distributed")
+pytest.importorskip("numpy")
+
+import numpy as np
+
+from distributed.utils_test import client  # noqa F401
+from distributed.utils_test import cluster_fixture  # noqa F401
+from distributed.utils_test import loop  # noqa F401
+
+import dask.array as da
+import dask.config
+from dask.array.utils import assert_eq
+
+
+def test_gh7632(client):
+    # Regression test for incorrect results due to SubgraphCallable.__eq__
+    # not correctly handling subgraphs with the same outputs and arity but
+    # different internals. The bug is triggered by distributed because it
+    # uses a function cache.
+    a = da.from_array(np.arange(3))
+    b = da.from_array(np.array([10 + 2j, 7 - 3j, 8 + 1j]))
+    cb = b.conj()
+    x = a * cb
+    (cb,) = dask.optimize(cb)
+    y = a * cb
+    expected = np.array([0 + 0j, 7 + 3j, 16 - 2j])
+    with dask.config.set({"optimization.fuse.active": False}):
+        x_value = x.compute()
+        y_value = y.compute()
+    assert_eq(x_value, expected)
+    assert_eq(y_value, expected)
```

Comment:
none of this has anything to do with the distributed scheduler; please move to the same test module that tests the rest of SubgraphCallable

Reply:
see my post in the discussion on how to test without using distributed

Reply:
Sure, I'll write some unit tests like this. I still think it's a good idea to have a black-box regression test, because it can catch future failure modes we haven't considered and because it's a usage pattern that isn't covered by the existing tests. I'll reframe it as "test that differently-optimised but equivalent collections work".